docs/webhooks/example.go

```diff
@@ -0,0 +1,149 @@
+// Copyright (c) 2022 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package webhooks provides example consumer code for Tailscale
+// webhooks.
+package webhooks
+
+import (
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+)
+
+type event struct {
+	Timestamp string            `json:"timestamp"`
+	Version   int               `json:"version"`
+	Type      string            `json:"type"`
+	Tailnet   string            `json:"tailnet"`
+	Message   string            `json:"message"`
+	Data      map[string]string `json:"data"`
+}
+
+const (
+	currentVersion = "v1"
+	secret         = "tskey-webhook-xxxxx" // sensitive, here just as an example
+)
+
+var (
+	errNotSigned     = errors.New("webhook has no signature")
+	errInvalidHeader = errors.New("webhook has an invalid signature")
+)
+
+func main() {
+	http.HandleFunc("/webhook", webhooksHandler)
+	if err := http.ListenAndServe(":80", nil); err != nil {
+		log.Fatal(err)
+	}
+}
+
+func webhooksHandler(w http.ResponseWriter, req *http.Request) {
+	defer req.Body.Close()
+	events, err := verifyWebhookSignature(req, secret)
+	if err != nil {
+		log.Printf("error validating signature: %v\n", err)
+	} else {
+		log.Printf("events received %v\n", events)
+		// Do something with your events. :)
+	}
+
+	// The handler should always report 2XX except in the case of
+	// transient failures (e.g. database backend is down).
+	// Otherwise your future events will be blocked by retries.
+}
+
+// verifyWebhookSignature checks the request's "Tailscale-Webhook-Signature"
+// header to verify that the events were signed by your webhook secret.
+// If verification fails, an error is reported.
+// If verification succeeds, the list of contained events is reported.
+func verifyWebhookSignature(req *http.Request, secret string) (events []event, err error) {
+	defer req.Body.Close()
+
+	// Grab the signature sent on the request header.
+	timestamp, signatures, err := parseSignatureHeader(req.Header.Get("Tailscale-Webhook-Signature"))
+	if err != nil {
+		return nil, err
+	}
+
+	// Verify that the timestamp is recent.
+	// Here, we use a threshold of 5 minutes.
+	if timestamp.Before(time.Now().Add(-time.Minute * 5)) {
+		return nil, fmt.Errorf("invalid header: timestamp older than 5 minutes")
+	}
+
+	// Form the expected signature.
+	b, err := io.ReadAll(req.Body)
+	if err != nil {
+		return nil, err
+	}
+	mac := hmac.New(sha256.New, []byte(secret))
+	mac.Write([]byte(fmt.Sprint(timestamp.Unix())))
+	mac.Write([]byte("."))
+	mac.Write(b)
+	want := hex.EncodeToString(mac.Sum(nil))
+
+	// Verify that the signatures match.
+	var match bool
+	for _, signature := range signatures[currentVersion] {
+		if signature == want {
```

Comment:
Almost certainly not a real issue, but as a best-practice, we should be using [`subtle.ConstantTimeCompare`](https://pkg.go.dev/crypto/subtle#ConstantTimeCompare) here.